src/geometry/algebra/lie_group.lean

```diff
@@ -0,0 +1,258 @@
+
+@[to_additive]
+lemma smooth_inv : smooth I I (λ x : G, x⁻¹) :=
+lie_group.smooth_inv
+
+@[to_additive]
+lemma smooth.inv {f : M → G}
+  (hf : smooth I' I f) : smooth I' I (λx, (f x)⁻¹) :=
+smooth_inv.comp hf
+
+@[to_additive]
+lemma smooth_on.inv {f : M → G} {s : set M}
+  (hf : smooth_on I' I f s) : smooth_on I' I (λx, (f x)⁻¹) s :=
+smooth_inv.comp_smooth_on hf
+
+end lie_group
+
+section prod_lie_group
+
+/- Instance of product group -/
+@[to_additive lie_add_group]
+instance {𝕜 : Type*} [nondiscrete_normed_field 𝕜] {H : Type*} [topological_space H]
+{E : Type*} [normed_group E] [normed_space 𝕜 E]  {I : model_with_corners 𝕜 E H}
+{G : Type*} [topological_space G] [charted_space H G] [group G] [topological_group G]
+[h : lie_group I G] {E' : Type*} [normed_group E'] [normed_space 𝕜 E']
+{H' : Type*} [topological_space H'] {I' : model_with_corners 𝕜 E' H'}
+{G' : Type*} [topological_space G'] [charted_space H' G']
+[group G'] [topological_group G'] [h' : lie_group I' G'] : lie_group (I.prod I') (G×G') :=
+{ smooth_mul := ((smooth_fst.comp smooth_fst).smooth.mul (smooth_fst.comp smooth_snd)).prod_mk
+    ((smooth_snd.comp smooth_fst).smooth.mul (smooth_snd.comp smooth_snd)),
+  smooth_inv := smooth_fst.inv.prod_mk smooth_snd.inv, }
+
+end prod_lie_group
+
+section lie_add_group_morphism
+
+variables {𝕜 : Type*} [nondiscrete_normed_field 𝕜]
+{E : Type*} [normed_group E] [normed_space 𝕜 E]
+{E' : Type*} [normed_group E'] [normed_space 𝕜 E']
+
+/-- Morphism of additive Lie groups. -/
+structure lie_add_group_morphism (I : model_with_corners 𝕜 E E) (I' : model_with_corners 𝕜 E' E')
+(G : Type*) [topological_space G] [charted_space E G] [smooth_manifold_with_corners I G]
+[add_group G] [topological_add_group G] [lie_add_group I G]
+(G' : Type*) [topological_space G'] [charted_space E' G'] [smooth_manifold_with_corners I' G']
+[add_group G'] [topological_add_group G'] [lie_add_group I' G'] extends add_monoid_hom G G' :=
+  (smooth_to_fun : smooth I I' to_fun)
+
+/-- Morphism of Lie groups. -/
+@[to_additive lie_add_group_morphism]
+structure lie_group_morphism (I : model_with_corners 𝕜 E E) (I' : model_with_corners 𝕜 E' E')
+(G : Type*) [topological_space G] [charted_space E G] [smooth_manifold_with_corners I G] [group G]
+[topological_group G] [lie_group I G]
+(G' : Type*) [topological_space G'] [charted_space E' G'] [smooth_manifold_with_corners I' G']
+[group G'] [topological_group G'] [lie_group I' G'] extends monoid_hom G G' :=
+  (smooth_to_fun : smooth I I' to_fun)
+
+variables {I : model_with_corners 𝕜 E E} {I' : model_with_corners 𝕜 E' E'}
+{G : Type*} [topological_space G] [charted_space E G] [smooth_manifold_with_corners I G]
+[group G] [topological_group G] [lie_group I G]
+{G' : Type*} [topological_space G'] [charted_space E' G'] [smooth_manifold_with_corners I' G']
+[group G'] [topological_group G'] [lie_group I' G']
+
+@[to_additive]
+instance : has_one (lie_group_morphism I I' G G') := ⟨⟨1, smooth_const⟩⟩
```

Comment:
I wouldn't call this `one`, because when `G = G'` I would instead expect `one` to be the identity morphism. It is definitely a good idea to define it, but with another name.

Reply:
`const`?

Reply:
Why is this case different than `monoid_hom`?

Reply:
Hmm, good point. I guess we want to use a type wrapper like `End` or `Aut` for the case `G = G'`. @sgouezel do you agree?

Reply:
It's funny how calling the zero morphism in the additive world `zero` feels natural to me, and calling the constant morphism equal to one by `one` feels strange. Anyway, I agree that since it's already like that for monoid homs we should either change both or keep it like that, and since you've already made a lot of things for this PR I'd say let's keep your `add_one` like that. We can change both later if someone insists strongly on it.